consensus/src/dag/reliable_broadcast.rs

```diff
@@ -184,3 +186,53 @@ impl RpcHandler for NodeBroadcastHandler {
         }
     }
 }
+
+#[derive(Debug, ThisError)]
+pub enum CertifiedNodeHandleError {
+    #[error("node already exists")]
+    NodeExists,
+    #[error("missing parents")]
+    MissingParents,
+}
+
+pub struct CertifiedNodeHandler {
+    dag: Arc<RwLock<Dag>>,
+}
+
+impl CertifiedNodeHandler {
+    pub fn new(dag: Arc<RwLock<Dag>>) -> Self {
+        Self { dag }
+    }
+}
+
+impl RpcHandler for CertifiedNodeHandler {
+    type Request = CertifiedNode;
+    type Response = CertifiedAck;
+
+    fn process(&mut self, node: Self::Request) -> anyhow::Result<Self::Response> {
+        {
+            let dag_reader = self.dag.read();
+            if dag_reader.exists(&node.digest()) {
+                bail!(CertifiedNodeHandleError::NodeExists);
```

Comment:
Why error? This can happen. Maybe trace/info?

Reply:
I don't think this is an error even, we probably need to send back ack still. imagine that a previous message triggered a node fetch and timeout the reliable broadcast but eventually successfully inserted. the second retry of reliable broadcast needs to receive ack